units/github_status_api.go

```diff
@@ -0,0 +1,320 @@
+		env: evergreen.GetEnvironment(),
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    githubStatusUpdateJobName,
+				Version: 0,
+				Format:  amboy.BSON,
+			},
+		},
+	}
+}
+
+// NewGithubStatusUpdateJobForBuild creates a job to update github's API from a Build.
+// Status will be reported as 'evergreen-[build variant name]'
+func NewGithubStatusUpdateJobForBuild(buildID string) amboy.Job {
+	job := makeGithubStatusUpdateJob()
+	job.FetchID = buildID
+	job.UpdateType = githubUpdateTypeBuild
+
+	job.SetID(fmt.Sprintf("%s:%s-%s-%s", githubStatusUpdateJobName, job.UpdateType, buildID, time.Now().String()))
+	return job
+}
+
+func repoReference(owner, repo string, prNumber int, ref string) string {
+	return fmt.Sprintf("%s/%s#%d@%s", owner, repo, prNumber, ref)
+}
+
+// NewGithubStatusUpdateForPatch creates a job to update github's API from a
+// Patch. Status will be reported as 'evergreen'
+func NewGithubStatusUpdateJobForPatch(version string) amboy.Job {
+	job := makeGithubStatusUpdateJob()
+	job.FetchID = version
+	job.UpdateType = githubUpdateTypePatch
+
+	job.SetID(fmt.Sprintf("%s:%s-%s-%s", githubStatusUpdateJobName, job.UpdateType, version, time.Now().String()))
+
+	return job
+}
+
+func (j *githubStatusUpdateJob) sendStatusUpdate(status *githubStatus) {
+	if !status.Valid() {
+		j.AddError(errors.New("status is invalid"))
+	}
+	if j.env.Settings() == nil || j.env.Settings().Ui.Url == "" {
+		j.AddError(errors.New("ui not configured"))
+	}
+	evergreenBaseURL := j.env.Settings().Ui.Url
+
+	githubOauthToken, err := j.env.Settings().GetGithubOauthToken()
+	if err != nil {
+		j.AddError(err)
+	}
+
+	token := strings.Split(githubOauthToken, " ")
+	if len(token) != 2 || token[0] != "token" {
+		j.AddError(errors.New("github token format expected to be 'token [oauthtoken]'"))
+	}
+	if j.HasErrors() {
+		return
+	}
+
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token[1]},
+	)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	tc := oauth2.NewClient(ctx, ts)
+	client := github.NewClient(tc)
+
+	newStatus := github.RepoStatus{
+		TargetURL:   github.String(fmt.Sprintf("%s%s", evergreenBaseURL, status.urlPath)),
+		Context:     github.String(status.context),
+		Description: github.String(status.description),
+		State:       github.String(status.ghStatus),
+	}
+	respStatus, resp, err := client.Repositories.CreateStatus(ctx, status.owner, status.repo, status.ref, &newStatus)
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+	resp.Body.Close()
+	if resp.StatusCode != http.StatusCreated {
+		repo := repoReference(status.owner, status.repo, status.prNumber, status.ref)
+		j.AddError(errors.Errorf("Github status creation for %s: expected http Status code: 201 Created, got %s", repo, http.StatusText(http.StatusCreated)))
+	}
+	if respStatus == nil {
+		j.AddError(errors.New("nil response from github"))
+	}
+}
+
+func (j *githubStatusUpdateJob) fetch(status *githubStatus) {
```

Comment:
I think methods *other* than Run should probably just return errors (and use grip.Catcher's as needed) rather than use AddError.